packages/fxa-auth-server/scripts/convert-customers-to-stripe-automatic-tax/helpers.ts

```diff
@@ -0,0 +1,100 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Stripe from 'stripe';
+
+/**
+ * Minimum number of days before monthly plan renewal to be upgraded to Stripe automatic tax
+ */
+const MONTHLY_NOTICE_DAYS = 14;
+/**
+ * Minimum number of days before 6-month and yearly renewal to be upgraded to Stripe automatic tax
+ */
+const YEARLY_NOTICE_DAYS = 30;
+/**
+ * Convert from Stripe dates to Unix milliseconds
+ */
+const SECONDS_TO_MILLISECONDS = 1000;
+
+/**
+ * Represents a mapping of user uids to IP addresses
+ */
+export interface IpAddressMap {
+  [key: string]: string;
+}
+
+/**
+ * Firestore subscriptions contain additional expanded information
+ * on top of the base Stripe.Subscription type
+ */
+export interface FirestoreSubscription extends Stripe.Subscription {
+  customer: string;
+  plan: Stripe.Plan;
+  price: Stripe.Price;
+}
+
+export class StripeAutomaticTaxConverterHelpers {
+  constructor() {}
+
+  /**
+   * Checks if Stripe customer is in a taxable location
+   * @returns True if the user is currently taxable
+   */
+  isTaxEligible(customer: Stripe.Customer) {
+    return customer.tax?.automatic_tax === 'supported';
+  }
+
+  /**
+   * Filters for all subscriptions that are eligible for upgrade to Stripe automatic tax
+   * @param subscriptions A list of subscriptions from Firestore
+   * @returns A filtered list of eligible subscriptions
+   */
+  filterEligibleSubscriptions(subscriptions: FirestoreSubscription[]) {
+    return subscriptions
+      .filter(this.willBeRenewed)
+      .filter(this.isStripeTaxDisabled)
+      .filter(this.isWithinNoticePeriod);
+  }
+
+  /**
+   * Check whether subscription is current & continuing
+   * @returns True if subscription is current & continuing
+   */
+  willBeRenewed(sub: Stripe.Subscription) {
+    return (
+      !sub.cancel_at && !sub.cancel_at_period_end && sub.status === 'active'
```

Comment:
If someone has themselves set to cancel at period end, then undoes it near the time it actually would happen, we'll end up skipping enabling tax on them?

Reply:
Fair point - I think my thought here is we don't want to send someone an email about their subscription rate changing if they've set themselves to cancel.

Though, if they remove their cancellation, we'd catch them in a future run of the script anyway.

Thoughts?

Reply:
I think that's fine, we could also later implement a 'heal' type scheme where if we see that they're not enabled on renewal we flag them somehow for an email and update.